processor/src/main/java/com/linecorp/decaton/processor/processors/BatchingProcessor.java

```diff
@@ -0,0 +1,137 @@
+            this.context = context;
+            this.task = task;
+        }
+    }
+
+    // visible for testing
+    BatchingProcessor(
+        long lingerMillis,
+        int capacity,
+        ScheduledThreadPoolExecutor scheduledExecutor
+    ) {
+        this.lingerMillis = lingerMillis;
+        this.capacity = capacity;
+
+        if (scheduledExecutor == null) {
+            scheduledExecutor = new ScheduledThreadPoolExecutor(1, r -> {
+                Thread th = new Thread(r);
+                th.setName("Decaton" + BatchingProcessor.class.getSimpleName()
+                           + '/' + System.identityHashCode(this));
+                return th;
+            });
+        }
+        // For faster shutdown cancel all pending flush on executor shutdown.
+        // In fact for this purpose we have two options here:
+        // A. Cancel all pending flush
+        // B. Run all pending flush immediately, regardless remaining time until flush
+        // A has a downside that it forces consumer to re-process some amount of tasks.
+        // B has a downside that it may causes flush bursting and then on spike for storage/API access in
+        // downstream processor.
+        // Given that this feature is expected to be used mainly for reducing workload for storage/APIs,
+        // we think taking A is much desirable behavior in most cases.
+        scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor = scheduledExecutor;
+    }
+
+    public BatchingProcessor(long lingerMillis) {
+        this(lingerMillis, Integer.MAX_VALUE, null);
+    }
+
+    public BatchingProcessor(long lingerMillis, int capacity) {
+        this(lingerMillis, capacity, null);
+    }
+
+    private void flush() {
+        if (!windowedTasks.isEmpty()) {
+            return;
+        }
+        synchronized (windowedTasks) {
+            processBatchingTasks(windowedTasks);
+        }
+    }
+
+    // visible for testing
+    Runnable flushTask() {
+        return this::flush;
+    }
+
+    private void scheduleFlush() {
+        executor.schedule(flushTask(), lingerMillis, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void process(ProcessingContext<T> context, T task) throws InterruptedException {
+        BatchingTask<T> newTask = new BatchingTask<>(context.deferCompletion(), context, task);
+        boolean isInitialTask = windowedTasks.isEmpty();
+        windowedTasks.add(newTask);
+        if (isInitialTask) {
+            scheduleFlush();
+        }
+        if (windowedTasks.size() >= this.capacity) {
+            flush();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        executor.shutdown();
+        executor.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+    }
+
+    /**
+     * *MUST* call {@link BatchingTask#completion}'s {@link DeferredCompletion#complete()} or
+     * {@link BatchingTask#context}'s {@link ProcessingContext#retry()} method.
```

Comment:
As a result of worrying about the retry process, I left it to the implementation side, but I think that it should be improved...

Reply:
Since this is the main method of this class that we expect users to implement, the document should be bit more informative IMO. 

At least it should cover topics like what to do after complete processing batch of tasks. (call `.completion().complete()` or  `retry()` for each right?),

what happens if an error is thrown, which threads may possibly call this method etc.